custom_metrics/responsive_images.js

```diff
@@ -0,0 +1,895 @@
+// takes <img> Element
+// returns a Number
+function totalNumberOfCandidates( img ) {
+
+	const candidateURLs = new Set();
+
+	if ( img.hasAttribute( 'src' ) ) {
+		candidateURLs.add( img.getAttribute( 'src' ) );
+	}
+
+	if ( img.hasAttribute( 'srcset' ) ) {
+		const parsedSrcset = parseSrcset( img.getAttribute( 'srcset' ) ).candidates;
+		parsedSrcset.forEach( candidate => {
+			candidateURLs.add( candidate.url );
+		} )
+	}
+
+	if ( img.parentNode.tagName === "PICTURE" ) {
+		const picture = img.parentNode,
+		      sources = pictureSources( picture );
+		sources.forEach( source => {
+			if ( source.hasAttribute( 'srcset' ) ) {
+				const parsedSrcset = parseSrcset( source.getAttribute( 'srcset' ) ).candidates;
+				parsedSrcset.forEach( candidate => {
+					candidateURLs.add( candidate.url );
+				} )
+			}
+		} );
+	}
+
+	candidateURLs.delete( '' ); // in case there are empties...
+	return candidateURLs.size;
+
+}
+
+// Is a <picture> doing media and/or type switching?
+// takes <picture> Element
+// returns {
+// 	mediaSwitching: Boolean,
+// 	typeSwitching: Boolean
+// }
+function pictureFeatures( picture ) {
+
+	let mediaSwitching = false,
+	    typeSwitching = false;
+
+	if ( picture.tagName !== "PICTURE" ) {
+		throw 'pictureFeatures() expected a <picture> Element as input but received something else'
+	}
+
+	const sources = pictureSources( picture );
+	sources.forEach( source => {
+		if ( source.hasAttribute( 'media' ) ) {
+			mediaSwitching = true;
+		}
+		if ( source.hasAttribute( 'type' ) ) {
+			typeSwitching = true;
+		}
+	} );
+
+	return { mediaSwitching, typeSwitching };
+
+}
+
+// Is a srcset using x or w descriptors?
+// takes a parsed srcset array (the `candidates` attribute from the ouput of the parseSrcset() function)
+// returns {
+// 	xDescriptors: Boolean,
+// 	wDescriptors: Boolean
+// }
+function srcsetFeatures( parsedSrcset ) {
+
+	const resourcesWithWDescriptors = parsedSrcset.filter( 
+		o => o.hasOwnProperty( 'w' )
+	);
+	const resourcesWithXDescriptors = parsedSrcset.filter( 
+		o => o.hasOwnProperty( 'd' )
+	);
+
+	return {
+		xDescriptors: resourcesWithXDescriptors.length > 0,
+		wDescriptors: resourcesWithWDescriptors.length > 0
+	}
+
+}
+
+// What features is an <img> using>?
+// takes an <img> Element
+// returns {
+// 	hasSrc: Boolean,
+// 	hasAlt: Boolean,
+// 	isInPicture:  Boolean,
+// 	hasCustomDataAttributes: Boolean,
+// 	hasWidthAndHeight: Boolean
+// }
+// we check for srcset and sizes use separately, in order to consider values supplied by the parent <picture> / <source> siblings
+// TODO consider loading (although there's already a metric for this?), decode,  crossorigin, referrerpolicy
+function imgFeatures( img ) {
+
+	return {
+		hasSrc: img.hasAttribute( 'src' ),
+		hasAlt: img.hasAttribute( 'alt' ),
+		isInPicture: img.parentNode.tagName === "PICTURE",
+		hasCustomDataAttributes: Object.keys( img.dataset ).length > 0,
+		hasWidthAndHeight: img.hasAttribute( 'width' ) &&
```

Comment:
Would it be useful to know if/how many images have either `width` or `height` but not both?

Reply:
Done